### PR TITLE
The ML.ENERGY Colosseum

### PR DESCRIPTION
**Pull request by jaywonchung**
I continued @AmberLJC's `serve` branch. This PR will override the original PR #21.

Please carefully review. With this amount of code, there has to be some bug.

## How to run the Gradio app for review

```console
$ cd leaderboard
$ git pull
$ git checkout jw-serve
$ pip install -e .[all]
$ COLOSSEUM_CONTROLLER_ADDR=[I will DM you this value] python app.py
```

## Changes from `serve`

List of notable changes.

Controller:
- `ExpiringDict` was not synchronizing the timestamp heap and the user dict on `__delitem__`. Fixed.
- The user prompt was sent to the controller in two separate POST requests, but the controller was lacking synchronization over the user state. Fixed.
  - Especially, randomly assigning model names could be called twice.
  - This was addressed by switching to async code (which serializes access to the user state dictionary). This also allows FastAPI to optimize better (according to the docs).
- User ID to Request ID
  - Users may play the Colosseum multiple times, and all those runs sharing a user ID simply makes things complicated. We did talk about finding malicious users through shared user IDs, but this detection method can just be avoided by refreshing anyway.
  - `X-User-ID` was removed. Now, `ControllerClient`, a client implementation that allows the Gradio app to talk to the backend controller, is wrapped in a `gr.State`. `ControllerClient.fork` copies the client and assigns a new request ID, which is called whenever the user submits a new prompt.
- `threading` to `asyncio.Task` -- The heartbeat task don't run every second of course, but sending a request to each worker involves across the network makes it a good candidate for an async IO task.
- Added incremental request state logging
  - Even if the user exits without voting, we still want to record at least part of the user's state (e.g., at least prompt, response, and energy, which will later turn out to be useful).
- Periodic heartbeat checks do not prevent a TGI server dying right after the check and the controller experiencing a connection error on `POST /generate_stream`.
  - The heartbeat task will periodically wake up and see if the dead TGI server is back up, and reinstate its worker state.

Gradio app:
- Added `queue=False` to event listeners that are not related to the Colosseum.
- Excessive code repetition due to the parallel structure of left and right votes, refactored.
- When the user clicks submit or presses enter, two parallel event handlers are launched. This made two things difficult: (1) removing the user's prompt from the input box (i.e., modifying the `TextBox` component) is a race condition that wasn't being handled (Until now there was only one user testing the app and the two initial handlers managed to both read the prompt, but when load increases, the handler that ran late may see an empty prompt text box because another handler deleted its content.) (2) enabling the response vote buttons only after both the models are done responding.
  - I consolidated generating the responses of two models into one event handler.
- The wordings of the final energy question changed. Now, it asks whether better response quality was worth the increase in energy consumption. I thought asking which model is more energy efficient could be confusing. I don't think there is a very obvious and unique definition of energy efficiency in our context.
- Autofocus the prompt textarea on reset.

Miscellaneous notes:
- I tried to see if I can write a Docker Compose file and deploy it with our Swarm cluster with `docker stack deploy`, but I wasn't able to get it working within a reasonable amount of time. GPUs being involved was the primary pain. So I just gave up and manually set up our stuff.

### CODE REVIEW

**Comment by jaywonchung** [comment]
Technically, I still need to handle stop_str. You can see that falcon-7b has a trailing word `User` in its response. I'll do that tomorrow. I posted the PR now since that's really a chunk of code inside the controllers' `prompt` function and won't affect any other place.

File: docs/colosseum_rules.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+> Enter the ML.ENERGY Colosseum, where language models duel with intellect, and your judgment tips the scales of victory.
+
+### Rules of the Colosseum
+
+- As the spectator, you'll decide the fates of two anonymous language models -- our gladiators.
+- Your role is twofold: First, you vote for the model that delivered the best response to your prompt.
+- Next, mighty [Zeus](https://ml.energy/zeus) will reveal which language model consumed more energy. Evaluate if its performance justified the energy consumption.
```
Comment (AmberLJC) [inline docs/colosseum_rules.md]: We may need to elaborate the measuring environment

File: docs/serving.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+docker swarm join
+```
+
+Create a custom swarm overlay network.
+
+```sh
+docker network create --attachable -d overlay leaderboard
+```
+
+Spawn the workers with `deployment/tgi-container.sh`
```
Comment (AmberLJC) [inline docs/serving.md]: where is this

File: docs/serving.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+docker_port=8000
+docker run --name=controller -v $HOME/leaderboard:/code  --network=mynetwork -p 8000:8000 -it  tgi:origin
+```
+
+```commandline
+python controller.py --host 0.0.0.0
+python app.py --share
+```
+
+The controller will check the live workers every min.
```
Comment (AmberLJC) [inline docs/serving.md]: clean up?

File: docs/serving.md (hunk 2 above)
Comment (jaywonchung) [reply docs/serving.md]: They're in our fork of text_generation_inference.

File: spitfight/colosseum/client.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+
+class ControllerClient:
+    """Client for the Colosseum controller, to be used by Gradio."""
+
+    def __init__(self, controller_addr: str, timeout: int = 15, request_id: UUID | None = None) -> None:
+        """Initialize the controller client."""
+        self.controller_addr = controller_addr
+        self.timeout = timeout
+        self.request_id = str(request_id) or str(uuid4())
```
Comment (AmberLJC) [inline spitfight/colosseum/client.py:32]: though i don't know when request_id would be None, `self.request_id = str(request_id) if request_id else str(uuid4())` could be better. when I manually run `test_new_uuid_on_deepcopy`, it gives error

File: spitfight/colosseum/client.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    if 400 <= response.status_code < 500:
+        raise gr.Error(response.json()["detail"])
+    elif response.status_code >= 500:
+        raise gr.Error("Failed to talk to our backend server. Please try again later.")
+
+
+class TestControllerClient(unittest.TestCase):
+    def test_new_uuid_on_deepcopy(self):
+        client = ControllerClient("http://localhost:8000")
+        clients = [deepcopy(client) for _ in range(50)]
```
Comment (AmberLJC) [inline spitfight/colosseum/client.py]: but when i change `deepcopy` to `client.fork()`, the unit test passed....

File: spitfight/colosseum/controller/worker.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        info = response.json()
+        if info["model_id"] != self.model_id:
+            raise ValueError(f"Model name mismatch: {info['model_id']} != {self.model_id}")
+        self.status = "up"
+
+    async def check_status(self) -> None:
+        """Check worker status and update `self.status` accordingly."""
+        async with httpx.AsyncClient() as client:
+            try:
+                response = await client.get(self.url + "/info")
```
Comment (AmberLJC) [inline spitfight/colosseum/controller/worker.py:62]: maybe minor, but why we choose '/info' not '/health'

File: app.py [hunk 7]
```diff
@@ -324,13 +335,260 @@ def plot_scatter(self, width, height, x, y, z):
...
+
+# Colosseum helper functions.
+def enable_interact():
+    return [gr.update(interactive=True)] * 2
+
+def disable_interact():
+    return [gr.update(interactive=False)] * 2
+
+def consumed_less_energy_message(energy_a, energy_b):
+    """Retuen a message that indicates that the user chose the model that consumed less energy.
```
Comment (AmberLJC) [inline app.py]: typo: return

**Review by AmberLJC — COMMENTED** [review]
Thanks for your great work! The code is very well-structured that I don't have too much to comment on. 
And for error handling stuff, I may need to look into them until we got enough feedbacks and collect some loggings. 
I may need to review the system metrics monitoring in the future as well.

File: spitfight/colosseum/client.py (hunk 4 above)
Comment (jaywonchung) [reply spitfight/colosseum/client.py:32]: My bad, that's old code. I used to rely on `gr.State`'s deepcopy behavior to assign new request IDs to the controller client, but now I just manually call `fork` inside the Gradio handler. I'll remove this part.

File: spitfight/colosseum/client.py (hunk 5 above)
Comment (jaywonchung) [reply spitfight/colosseum/client.py]: Yeah exactly haha. Sorry about the confusion.

File: spitfight/colosseum/controller/worker.py (hunk 6 above)
Comment (jaywonchung) [reply spitfight/colosseum/controller/worker.py:62]: That is to check whether `model_id` returned by `GET /info` matches `self.model_id`.

File: docs/colosseum_rules.md [hunk 8]
```diff
@@ -0,0 +1,9 @@
+> Enter the ML.ENERGY Colosseum, where language models duel with intellect, and your judgment tips the scales of victory.
+
```
Comment (AmberLJC) [inline docs/colosseum_rules.md]: we should include our social media links or any ways that can interact with users

Comment (jaywonchung) [reply docs/colosseum_rules.md]: Good point. I don't think it deserves a space at the top of the page, so let me extend `colosseum_terms.md` with a link to our repo's discussion board. That way users can ask questions and report problems.

File: spitfight/colosseum/controller/worker.py (hunk 6 above)
Comment (AmberLJC) [reply spitfight/colosseum/controller/worker.py:62]: (minor) but do we just need to do that in `audit` and the model_id would be fine afterwards

Comment (jaywonchung) [reply spitfight/colosseum/controller/worker.py:62]: I see your point. But when a worker dies for some reason, currently we have to manually restart the worker, and we might make the mistake of starting the worker with another model. This check will guard against that mistake, so I'd rather leave this as is.

**Comment by jaywonchung** [comment]
@AmberLJC Can you take a quick look at the `stop_str` handling part in `Controller.prompt` in `controller.py`? I tested it for Falcon (`stop_str="\nUser"`) and Llama2 (No `stop_str`) and it seems fine for both. I'll address all of your comments later today and let's proceed to deployment and internal testing after that.

File: docs/colosseum_rules.md (hunk 1 above)
Comment (jaywonchung) [reply docs/colosseum_rules.md]: Note for myself: Should also talk about max num tokens

**Comment by AmberLJC** [comment]
> @AmberLJC Can you take a quick look at the `stop_str` handling part in `Controller.prompt` in `controller.py`? I tested it for Falcon (`stop_str="\nUser"`) and Llama2 (No `stop_str`) and it seems fine for both. I'll address all of your comments later today and let's proceed to deployment and internal testing after that.

I think it works perfect on my side.
But curious, why fastchat do partial [check](https://github.com/lm-sys/FastChat/blob/32c1910ebb3567545556f9cb859ba02b36b35ed6/fastchat/serve/inference.py#L207)

**Comment by jaywonchung** [comment]
Thanks for testing out!

> But curious, why fastchat do partial [check](https://github.com/lm-sys/FastChat/blob/32c1910ebb3567545556f9cb859ba02b36b35ed6/fastchat/serve/inference.py#L207)

We do it, too. That the part of the code that deals with 'partial match'.

For instance, for Falcon, `stop_str` is `\nUser`, which is at least two tokens. Now, say our `client.generate_stream` gives us `\n`. What do we do?

One way is to just yield `\n`. But what if that was actually part of `stop_str` and the next token is `User`? Then the model would have generated the `stop_str`, which we don't want to give to the user. (Note that yielding `\n` seems fine for this particular case, but model `stop_str`s can be pretty random depending on the model, like `Human:`).

That's why we should keep track of partial matches for `stop_str`. If the token we received from TGI is a *prefix* of `stop_str`, instead of giving it to the user, we buffer it in `buffer`, until we know whether this is going to become a full match of `stop_str` or not.

**Pull request merged**